contracts/layer-2/TokenTransferRelay.sol

```diff
@@ -0,0 +1,292 @@
+    {
+        return
+            interfaceId == type(IERC721Receiver).interfaceId ||
+            AccessControl.supportsInterface(interfaceId);
+    }
+
+    /**
+     * @dev See {IERC721Receiver-onERC721Received}.
+     */
+    function onERC721Received(
+        address operator,
+        address from,
+        uint256 tokenId,
+        bytes calldata data
+    ) external pure returns (bytes4) {
+        return this.onERC721Received.selector;
+    }
+
+    /**
+     * @dev sets the address to which ERC20 tokens should be sent to.
+     * The function caller must have been granted MAINTENANCE_ROLE.
+     */
+    function setERC721Receiver(address _erc721Receiver)
+        external
+        onlyRole(MAINTENANCE_ROLE)
+    {
+        ERC721Receiver = _erc721Receiver;
+    }
+
+    /**
+     * @dev sets the address to which ERC721 tokens should be sent to.
+     * The function caller must have been granted MAINTENANCE_ROLE.
+     */
+    function setERC20Receiver(address _erc20Receiver)
+        external
+        onlyRole(MAINTENANCE_ROLE)
+    {
+        ERC20Receiver = _erc20Receiver;
+    }
+
+    function getTransferKey(bytes32 UID, address from)
+        public
+        pure
+        returns (bytes32)
+    {
+        return keccak256(abi.encodePacked(UID, from));
+    }
+
+    function getTransfer(bytes32 UID, address from)
+        public
+        view
+        returns (Transfer memory)
+    {
+        return _transfers[getTransferKey(UID, from)];
+    }
+
+    function isTransferReserved(bytes32 UID, address from)
+        public
+        view
+        returns (bool)
+    {
+        return getTransfer(UID, from).state == TRANSFER_RESERVED;
+    }
+
+    function isTransferProcessed(bytes32 UID, address from)
+        public
+        view
+        returns (bool)
+    {
+        bytes32 state = getTransfer(UID, from).state;
+        return state == TRANSFER_EXECUTED || state == TRANSFER_REVERTED;
+    }
+
+    function reserveTransfer(
+        bytes32 UID,
+        uint256[] calldata tokenIds,
+        uint256 amount
+    ) external {
+        _reserveTransfer(UID, msg.sender, tokenIds, amount);
+    }
+
+    function reserveTransferFrom(
+        bytes32 UID,
+        address from,
+        uint256[] calldata tokenIds,
+        uint256 amount
+    ) external onlyRole(OPERATOR_ROLE) {
+        _reserveTransfer(UID, from, tokenIds, amount);
+    }
+
+    function executeTransfer(bytes32 UID) external {
+        _executeTransfer(UID, msg.sender);
+    }
+
+    function executeTransferFrom(bytes32 UID, address from)
+        external
+        onlyRole(OPERATOR_ROLE)
+    {
+        _executeTransfer(UID, from);
+    }
+
+    function revertTransfer(bytes32 UID, address from)
+        external
+        onlyRole(OPERATOR_ROLE)
+    {
+        require(
+            isTransferReserved(UID, from),
+            "TokenTransferRelay: Transfer is not reserved"
+        );
+
+        Transfer storage transfer = _transfers[getTransferKey(UID, from)];
```

Comment:
Why not call `getTransfer` method ?

Reply:
getTransfer returns a `memory` pointer, which is not convertible to a `storage` type pointer.

However, it might be worth it to declare this variable to `memory` due to the multiple `transfer.*` accesses along the function, which would be cheaper from memory. However, we'd have to deal with the cost of loading the entire struct to memory first. I'll run some gas cost tests and leave the conclusions here.

Note also that the `transfer.tokenIds` array is loaded into memory when calling `_batchERC721Transfer`. Since we need to read from a new slot of the array in every loop iteration, this makes running that function potentially orders of magnitude cheaper, as opposed to reading directly from storage.